File: src/templates/skills/bbva-brand/scripts/ts/generate_docx.ts
#!/usr/bin/env npx tsx
/**
 * generate_docx.ts — BBVA-branded DOCX generator (docx npm, DEFAULT runtime).
 *
 * Usage:
 *   npx tsx generate_docx.ts --content content.json --output output.docx
 *
 * content.json schema: { title, subtitle?, author?, sections: [{number, label, heading, body, items?, callout?}] }
 *
 * Install dependency: npm install docx
 */
import {
  Document,
  Packer,
  Paragraph,
  TextRun,
  HeadingLevel,
  AlignmentType,
  BorderStyle,
  ShadingType,
  PageBreak,
  UnderlineType,
  convertInchesToTwip,
} from "docx";
import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import * as bt from "./brand_tokens.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface Section {
  number: string;
  label: string;
  heading: string;
  body: string;
  items?: string[];
  callout?: string;
}

interface Content {
  title: string;
  subtitle?: string;
  author?: string;
  sections: Section[];
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function hexToDocx(key: string): string {
  return bt.hex(key).toUpperCase();
}

function coverPage(content: Content): Paragraph[] {
  const paras: Paragraph[] = [];

  paras.push(
    new Paragraph({
      children: [
        new TextRun({
          text: "BBVA",
          bold: true,
          size: 36,
          color: hexToDocx("primary"),
          font: bt.FONTS["pptx_body"],
        }),
      ],
      spacing: { after: 400 },
    })
  );

  paras.push(
    new Paragraph({
      children: [
        new TextRun({
          text: content.title,
          bold: true,
          size: 56,
          color: hexToDocx("primary_dark"),
          font: bt.FONTS["pptx_headlines"],
        }),
      ],
      heading: HeadingLevel.TITLE,
      spacing: { after: 240 },
    })
  );

  if (content.subtitle) {
    paras.push(
      new Paragraph({
        children: [
          new TextRun({
            text: content.subtitle,
            size: 28,
            italics: true,
            color: hexToDocx("text_secondary"),
            font: bt.FONTS["pptx_body"],
          }),
        ],
        spacing: { after: 480 },
      })
    );
  }

  if (content.author) {
    paras.push(
      new Paragraph({
        children: [
          new TextRun({
            text: content.author,
            size: 22,
            color: hexToDocx("text_light"),
            font: bt.FONTS["pptx_body"],
          }),
        ],
        spacing: { after: 120 },
      })
    );
  }

  paras.push(
    new Paragraph({
      children: [
        new TextRun({
          text: new Date().toLocaleDateString("es-ES", { year: "numeric", month: "long" }),
          size: 20,
          color: hexToDocx("text_light"),
          font: bt.FONTS["pptx_body"],
        }),
      ],
      spacing: { after: 2400 },
    })
  );

  // Page break after cover
  paras.push(new Paragraph({ children: [new PageBreak()] }));

  return paras;
}

function sectionDivider(section: Section): Paragraph[] {
  return [
    new Paragraph({
      children: [
        new TextRun({
          text: `${section.number}  ${section.label.toUpperCase()}`,
          bold: true,
          size: 24,
          color: hexToDocx("primary"),
          font: bt.FONTS["pptx_body"],
          characterSpacing: 60,
        }),
      ],
      border: {
        bottom: { color: hexToDocx("primary"), size: 6, space: 1, style: BorderStyle.SINGLE },
      },
      spacing: { before: 600, after: 200 },
    }),
    new Paragraph({
      children: [
        new TextRun({
          text: section.heading,
          bold: true,
          size: 40,
          color: hexToDocx("primary_dark"),
          font: bt.FONTS["pptx_headlines"],
        }),
      ],
      heading: HeadingLevel.HEADING_1,
      spacing: { after: 300 },
    }),
  ];
}

function sectionBody(section: Section): Paragraph[] {
  const paras: Paragraph[] = [];

  paras.push(
    new Paragraph({
      children: [
        new TextRun({
          text: section.body,
          size: 24,
          color: hexToDocx("text_primary"),
          font: bt.FONTS["pptx_body"],
        }),
      ],
      spacing: { after: 240, line: 360, lineRule: "auto" },
    })
  );

  if (section.items && section.items.length > 0) {
    for (const item of section.items) {
      paras.push(
        new Paragraph({
          children: [
            new TextRun({
              text: item,
              size: 24,
              color: hexToDocx("text_primary"),
              font: bt.FONTS["pptx_body"],
            }),
          ],
          bullet: { level: 0 },
          spacing: { after: 120 },
        })
      );
    }
  }

  if (section.callout) {
    paras.push(
      new Paragraph({
        children: [
          new TextRun({
            text: section.callout,
            size: 24,
            italics: true,
            color: hexToDocx("primary_dark"),
            font: bt.FONTS["pptx_body"],
          }),
        ],
        shading: {
          type: ShadingType.SOLID,
          color: "E8EBF7",
          fill: "E8EBF7",
        },
        border: {
          left: { color: hexToDocx("primary"), size: 12, space: 4, style: BorderStyle.SINGLE },
        },
        indent: { left: convertInchesToTwip(0.2), right: convertInchesToTwip(0.2) },
        spacing: { before: 240, after: 240 },
      })
    );
  }

  return paras;
}

// ---------------------------------------------------------------------------
// Main generator
// ---------------------------------------------------------------------------

export async function generateDocx(content: Content, outputPath: string): Promise<void> {
  const allParagraphs: Paragraph[] = [
    ...coverPage(content),
  ];

  for (const section of content.sections) {
    allParagraphs.push(...sectionDivider(section));
    allParagraphs.push(...sectionBody(section));
  }

  const doc = new Document({
    title: content.title,
    subject: content.subtitle ?? "",
    creator: content.author ?? "BBVA",
    sections: [
      {
        properties: {
          page: {
            margin: {
              top: convertInchesToTwip(1),
              bottom: convertInchesToTwip(1),
              left: convertInchesToTwip(1.2),
              right: convertInchesToTwip(1.2),
            },
          },
        },
        children: allParagraphs,
      },
    ],
  });

  const buffer = await Packer.toBuffer(doc);
  writeFileSync(outputPath, buffer);
  console.log(`DOCX written: ${outputPath} (${content.sections.length} sections)`);
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

function parseCli(): { content: Content; output: string } {
  const args = process.argv.slice(2);
  const getArg = (flag: string): string | undefined => {
    const idx = args.indexOf(flag);
    return idx !== -1 ? args[idx + 1] : undefined;
  };

  const contentFile = getArg("--content");
  const output = getArg("--output") ?? "output.docx";

  if (contentFile) {
    const raw = JSON.parse(readFileSync(contentFile, "utf-8")) as Content;
    return { content: raw, output };
  }

  throw new Error("Usage: npx tsx generate_docx.ts --content content.json --output output.docx");
}

if (process.argv[1] && fileURLToPath(import.meta.url).endsWith(process.argv[1].split("/").pop()!)) {
  const { content, output } = parseCli();
  generateDocx(content, output).catch((err: unknown) => {
    console.error("DOCX generation failed:", err);
    process.exit(1);
  });
}

File: src/templates/skills/bbva-brand/scripts/python/generate_docx.py
#!/usr/bin/env python3
"""
generate_docx.py — BBVA-branded DOCX generator (python-docx, FALLBACK runtime).

Usage:
    python3 generate_docx.py --content content.json --output output.docx

Install dependency: pip install python-docx
content.json schema: {title, subtitle?, author?, sections: [{number, label, heading, body, items?, callout?}]}
"""

import sys
import json
import argparse
from pathlib import Path
from datetime import datetime

sys.path.insert(0, str(Path(__file__).parent))
import brand_tokens as bt

from docx import Document
from docx.shared import Pt, Cm, RGBColor
from docx.oxml.ns import qn
from docx.oxml import OxmlElement


def _rgb(key: str) -> RGBColor:
    r, g, b = bt.rgb(key)
    return RGBColor(r, g, b)


def _set_font(
    run,
    color_key: str,
    size_pt: int,
    font_key: str = "pptx_body",
    bold=False,
    italic=False,
) -> None:
    run.font.color.rgb = _rgb(color_key)
    run.font.size = Pt(size_pt)
    run.font.name = bt.FONTS.get(font_key, "Arial")
    run.font.bold = bold
    run.font.italic = italic


def _add_callout(doc: Document, text: str) -> None:
    para = doc.add_paragraph()
    run = para.add_run(text)
    _set_font(run, "primary_dark", 12, italic=True)

    # Add left border via direct XML (python-docx lacks native paragraph borders)
    pPr = para._p.get_or_add_pPr()
    pBdr = OxmlElement("w:pBdr")
    left = OxmlElement("w:left")
    left.set(qn("w:val"), "single")
    left.set(qn("w:sz"), "12")
    left.set(qn("w:space"), "4")
    left.set(qn("w:color"), bt.hex_color("primary").upper())
    pBdr.append(left)
    pPr.append(pBdr)

    # Shading
    shd = OxmlElement("w:shd")
    shd.set(qn("w:val"), "clear")
    shd.set(qn("w:fill"), "E8EBF7")
    pPr.append(shd)

    para.paragraph_format.space_before = Pt(12)
    para.paragraph_format.space_after = Pt(12)
    para.paragraph_format.left_indent = Cm(0.5)


def build_cover(doc: Document, content: dict) -> None:
    # BBVA wordmark
    p = doc.add_paragraph()
    run = p.add_run("BBVA")
    _set_font(run, "primary", 18, bold=True)
    p.paragraph_format.space_after = Pt(20)

    # Title
    p = doc.add_paragraph()
    run = p.add_run(content["title"])
    _set_font(run, "primary_dark", 28, font_key="pptx_headlines", bold=True)
    p.paragraph_format.space_after = Pt(12)

    # Subtitle
    if content.get("subtitle"):
        p = doc.add_paragraph()
        run = p.add_run(content["subtitle"])
        _set_font(run, "text_secondary", 14, italic=True)
        p.paragraph_format.space_after = Pt(24)

    # Author + date
    meta_parts = []
    if content.get("author"):
        meta_parts.append(content["author"])
    meta_parts.append(datetime.now().strftime("%B %Y"))
    p = doc.add_paragraph()
    run = p.add_run(" · ".join(meta_parts))
    _set_font(run, "text_light", 11)
    p.paragraph_format.space_after = Pt(36)

    doc.add_page_break()


def build_section(doc: Document, section: dict) -> None:
    # Section number + label (heading 2)
    p = doc.add_paragraph()
    run = p.add_run(f"{section['number']}  {section['label'].upper()}")
    _set_font(run, "primary", 11, bold=True)
    p.paragraph_format.space_before = Pt(30)
    p.paragraph_format.space_after = Pt(6)

    # Horizontal rule (via bottom border on paragraph)
    pPr = p._p.get_or_add_pPr()
    pBdr = OxmlElement("w:pBdr")
    bottom = OxmlElement("w:bottom")
    bottom.set(qn("w:val"), "single")
    bottom.set(qn("w:sz"), "6")
    bottom.set(qn("w:space"), "1")
    bottom.set(qn("w:color"), bt.hex_color("primary").upper())
    pBdr.append(bottom)
    pPr.append(pBdr)

    # Section heading (heading 1)
    heading = doc.add_heading(section["heading"], level=1)
    for run in heading.runs:
        _set_font(run, "primary_dark", 20, font_key="pptx_headlines", bold=True)
    heading.paragraph_format.space_after = Pt(12)

    # Body
    p = doc.add_paragraph(section["body"])
    p.paragraph_format.space_after = Pt(8)
    p.paragraph_format.line_spacing = Pt(18)
    for run in p.runs:
        _set_font(run, "text_primary", 12)

    # Bullet items
    for item in section.get("items", []):
        p = doc.add_paragraph(item, style="List Bullet")
        for run in p.runs:
            _set_font(run, "text_primary", 12)
        p.paragraph_format.space_after = Pt(4)

    # Callout
    if section.get("callout"):
        _add_callout(doc, section["callout"])


def generate_docx(content: dict, output_path: str) -> None:
    doc = Document()

    # Page margins
    for section in doc.sections:
        section.top_margin = Cm(2.5)
        section.bottom_margin = Cm(2.5)
        section.left_margin = Cm(3)
        section.right_margin = Cm(3)

    build_cover(doc, content)

    for sec in content.get("sections", []):
        build_section(doc, sec)

    doc.save(output_path)
    n = len(content.get("sections", []))
    print(f"DOCX written: {output_path} ({n} sections)")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="BBVA DOCX generator (python-docx fallback)"
    )
    parser.add_argument("--content", required=True, help="Path to content.json")
    parser.add_argument("--output", default="output.docx", help="Output .docx file")
    args = parser.parse_args()

    with open(args.content, "r", encoding="utf-8") as f:
        content = json.load(f)

    generate_docx(content, args.output)


if __name__ == "__main__":
    main()
